### PR TITLE
[RAPTOR-5491] fix datatypes in

### PR DESCRIPTION
**Pull request by scottp-ml**
# This repository is public. Do not put here any private DataRobot or customer's data: code, datasets, model artifacts, .etc.

## Summary
This fixes the use of IN to match DR

### CODE REVIEW

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 1]
```diff
@@ -265,16 +263,16 @@ def validate(self, dataframe):
         base_error = f"Datatypes incorrect. Data has types: {DataTypes.list_str(types_present)}"
 
         errors = {
-            Conditions.EQUALS: f"{base_error}, but expected only {self.values[0]}.",
+            Conditions.EQUALS: f"{base_error}, but expected types to exactly match: {DataTypes.list_str(self.values)}",
             Conditions.NOT_EQUALS: f"{base_error}, but expected {self.values[0]} to NOT be present.",
-            Conditions.IN: f"{base_error}, but expected types to exactly match: {DataTypes.list_str(self.values)}",
+            Conditions.IN: f"{base_error}, but expected only {DataTypes.list_str(self.values)}.",
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:268]: maybe `which include values that are not in {}`

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 2]
```diff
@@ -265,16 +263,16 @@ def validate(self, dataframe):
...
             Conditions.NOT_IN: f"{base_error}, but expected no types in: {DataTypes.list_str(self.values)} to be present",
         }
 
         tests = {
-            Conditions.EQUALS: lambda data_types: self.values == data_types,
+            Conditions.EQUALS: lambda data_types: set(self.values) == set(data_types),
             Conditions.NOT_EQUALS: lambda data_types: self.values[0] not in data_types,
-            Conditions.IN: lambda data_types: set(self.values) == set(data_types),
+            Conditions.IN: lambda data_types: len(set(data_types) - set(self.values)) == 0,
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:275]: is this supposed to be a subset?

Comment (4dahalibut) [reply custom_model_runner/datarobot_drum/drum/typeschema_validation.py:275]: If so, please just use `.issubset`

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 3]
```diff
@@ -265,16 +263,16 @@ def validate(self, dataframe):
         base_error = f"Datatypes incorrect. Data has types: {DataTypes.list_str(types_present)}"
 
         errors = {
-            Conditions.EQUALS: f"{base_error}, but expected only {self.values[0]}.",
+            Conditions.EQUALS: f"{base_error}, but expected types to exactly match: {DataTypes.list_str(self.values)}",
             Conditions.NOT_EQUALS: f"{base_error}, but expected {self.values[0]} to NOT be present.",
-            Conditions.IN: f"{base_error}, but expected types to exactly match: {DataTypes.list_str(self.values)}",
+            Conditions.IN: f"{base_error}, which include values that are not in  {DataTypes.list_str(self.values)}.",
```
Comment (4dahalibut) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:268]: includes*

**Review by 4dahalibut — APPROVED** [review]

**Review by jmbannon — APPROVED** [review]

File: custom_model_runner/datarobot_drum/drum/typeschema_validation.py [hunk 4]
```diff
@@ -265,16 +263,16 @@ def validate(self, dataframe):
...
+            Conditions.EQUALS: f"{base_error}, but expected types to exactly match: {DataTypes.list_str(self.values)}",
             Conditions.NOT_EQUALS: f"{base_error}, but expected {self.values[0]} to NOT be present.",
-            Conditions.IN: f"{base_error}, but expected types to exactly match: {DataTypes.list_str(self.values)}",
+            Conditions.IN: f"{base_error}, which includes values that are not in {DataTypes.list_str(self.values)}.",
             Conditions.NOT_IN: f"{base_error}, but expected no types in: {DataTypes.list_str(self.values)} to be present",
         }
 
         tests = {
-            Conditions.EQUALS: lambda data_types: self.values == data_types,
+            Conditions.EQUALS: lambda data_types: set(self.values) == set(data_types),
```
Comment (eric-s-s) [inline custom_model_runner/datarobot_drum/drum/typeschema_validation.py:273]: do we allow multiple data types for equals?

**Pull request merged**